File: src/services/metricsService.js
// Productivity metrics and scoring service
import { dataService } from './dataService';

export const metricsService = {
  // Calculate daily productivity score
  calculateDailyScore: (habits, goals, scheduleCompletion) => {
    if (!habits) return 0;

    const habitCount = Object.keys(habits).length;
    const completedHabits = Object.values(habits).filter(Boolean).length;
    const habitScore = habitCount > 0 ? (completedHabits / habitCount) * 50 : 0;
    
    const scheduleScore = scheduleCompletion * 30;
    const goalProgressScore = goals.reduce((acc, goal) => {
      const progress = goal.current / goal.target;
      return acc + (progress * 20 / goals.length);
    }, 0);

    return Math.round(habitScore + scheduleScore + goalProgressScore);
  },

  // Calculate streak for a habit
  calculateStreak: (userId, habitKey, history) => {
    if (!history || history.length === 0) return 0;

    let streak = 0;
    const today = new Date().toISOString().split('T')[0];
    
    // Check today first
    const todayData = history.find(h => h.date === today);
    if (todayData && todayData.habits && todayData.habits[habitKey]) {
      streak = 1;
    } else {
      return 0; // No streak if today is not completed
    }

    // Count backwards
    for (let i = 1; i < history.length; i++) {
      const date = new Date();
      date.setDate(date.getDate() - i);
      const dateStr = date.toISOString().split('T')[0];
      const dayData = history.find(h => h.date === dateStr);
      
      if (dayData && dayData.habits && dayData.habits[habitKey]) {
        streak++;
      } else {
        break;
      }
    }

    return streak;
  },

  // Get insights (what went well, what to improve)
  generateInsights: (userId, days = 7) => {
    const history = dataService.getHabitHistory(userId, days);
    const metricsHistory = dataService.getMetricsHistory(userId, days);
    
    if (history.length === 0) {
      return {
        strengths: [],
        improvements: [],
        trends: [],
      };
    }

    // Analyze habit completion rates
    const habitStats = {};
    history.forEach(day => {
      if (day.habits) {
        Object.keys(day.habits).forEach(habitKey => {
          if (!habitStats[habitKey]) {
            habitStats[habitKey] = { completed: 0, total: 0 };
          }
          habitStats[habitKey].total++;
          if (day.habits[habitKey]) {
            habitStats[habitKey].completed++;
          }
        });
      }
    });

    // Find strengths (habits with >70% completion)
    const strengths = Object.entries(habitStats)
      .filter(([_, stats]) => stats.total > 0 && (stats.completed / stats.total) > 0.7)
      .map(([habitKey, stats]) => ({
        habit: habitKey,
        completionRate: Math.round((stats.completed / stats.total) * 100),
      }))
      .sort((a, b) => b.completionRate - a.completionRate);

    // Find improvements (habits with <50% completion)
    const improvements = Object.entries(habitStats)
      .filter(([_, stats]) => stats.total > 0 && (stats.completed / stats.total) < 0.5)
      .map(([habitKey, stats]) => ({
        habit: habitKey,
        completionRate: Math.round((stats.completed / stats.total) * 100),
      }))
      .sort((a, b) => a.completionRate - b.completionRate);

    // Calculate trend (improving or declining)
    const scores = metricsHistory.map(m => m.score || 0);
    const trend = scores.length >= 2 
      ? (scores[0] > scores[scores.length - 1] ? 'improving' : 'declining')
      : 'stable';

    return {
      strengths,
      improvements,
      trends: {
        direction: trend,
        averageScore: scores.length > 0 
          ? Math.round(scores.reduce((a, b) => a + b, 0) / scores.length)
          : 0,
        bestDay: metricsHistory.length > 0
          ? metricsHistory.reduce((best, current) => 
              (current.score || 0) > (best.score || 0) ? current : best
            )
          : null,
      },
    };
  },

  // Get productivity score for a date range
  getProductivityScore: (userId, days = 7) => {
    const metricsHistory = dataService.getMetricsHistory(userId, days);
    if (metricsHistory.length === 0) return 0;

    const scores = metricsHistory.map(m => m.score || 0);
    return Math.round(scores.reduce((a, b) => a + b, 0) / scores.length);
  },
};

File: backend/ml/metrics_service.py
"""
Metrics calculation service
"""
from typing import List, Dict
from datetime import datetime

class MetricsService:
    """Calculate productivity metrics and scores"""
    
    @staticmethod
    def calculate_daily_score(habits: Dict[str, bool], goals: List[Dict], schedule_completion: float = 0.8) -> float:
        """Calculate daily productivity score (0-100)"""
        if not habits:
            return 0.0
        
        habit_count = len(habits)
        completed_habits = sum(1 for v in habits.values() if v)
        habit_score = (completed_habits / habit_count) * 50 if habit_count > 0 else 0
        
        schedule_score = schedule_completion * 30
        
        goal_progress_score = 0
        if goals:
            total_progress = sum(
                (g.get('current_value', 0) / g.get('target_value', 1)) * (20 / len(goals))
                for g in goals if g.get('target_value', 0) > 0
            )
            goal_progress_score = min(total_progress, 20)
        
        return round(habit_score + schedule_score + goal_progress_score, 2)
    
    @staticmethod
    def calculate_streak(habit_history: List[Dict], habit_key: str) -> int:
        """Calculate streak for a specific habit"""
        if not habit_history:
            return 0
        
        # Sort by date (most recent first)
        sorted_history = sorted(habit_history, key=lambda x: x.get('date', ''), reverse=True)
        
        streak = 0
        today = datetime.now().strftime("%Y-%m-%d")
        
        # Check today
        today_data = next((h for h in sorted_history if h.get('date') == today), None)
        if today_data and today_data.get('habits', {}).get(habit_key):
            streak = 1
        else:
            return 0
        
        # Count backwards
        for i, day_data in enumerate(sorted_history[1:], 1):
            if day_data.get('habits', {}).get(habit_key):
                streak += 1
            else:
                break
        
        return streak

